### ISSUE TITLE
Add support for recording baseMetrics

### ISSUE THREAD

**Issue by joemcgill** [10up/wordpress-develop#31]
### Is your enhancement related to a problem? Please describe.

After discussion with @youknowriad, it seems that some of our assumptions were wrong about how `baseMetrics` are used when displaying metrics on the [codehealth dashboard](https://codehealth.vercel.app/).

To account for possible variance in the data introduced by changes to the GitHub runner environments, each metric that gets displayed in the graph is first normalized against a reference commit using a `baseMetric` value that is recorded from a consistent commit or `baseHash` (e.g., the last major release). These `bashMetrics` are then used to normalize all other metrics using the following formula before a metric is recorded:

```
metricValue * baseMetricValue (original) / baseMetricValue (current run)
```

## Acceptance Criteria

- During each performance workflow, record the base performance metrics for the previous major release hash (WP 6.1.1 i.e., 031cedc06d612c4a74f7b6d3e77d8145f2954a25)
- Update the `log-results.js` script to include the correct baseMetrics and baseHash to ensure normalization is happening correctly.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by eclarke1** [comment]
@joemcgill would it be possible to please add an estimate here for the work, using the labels?

**Comment by felixarntz** [comment]
@joemcgill ACs LGTM!

Maybe we can annotate with a comment what this base commit hash refers to and also that it should not be changed?

It would be good to engage separately in regards to how Gutenberg does it: From my perspective, changing the hash with every version is not a good idea as it makes performance comparisons across versions inaccurate and thus more or less worthless. It's still useful then for checking performance within one release, but not for holistic metrics from version to version. Maybe that's the point, but it certainly is an important caveat to clarify with anyone who uses these dashboards.

**Comment by joemcgill** [comment]
> From my perspective, changing the hash with every version is not a good idea as it makes performance comparisons across versions inaccurate

I tend to agree with you, @felixarntz, based on my understanding of how the dashboard is normalizing data based on these baseMetrics. I think this is something we can open a discussion on [in the repo](https://github.com/youknowriad/codehealth) for that dashboard.

**Comment by mukeshpanchal27** [comment]
Thanks @joemcgill for issue. I have some questions.
- If we measure the previous WP version metric, do we need to follow [Gutenberg approach](https://github.com/WordPress/gutenberg/blob/trunk/.github/workflows/performance.yml#L78-L87)? When I go over it, it's the most difficult part.
- Is it really necessary to measure previous WordPress version metrics on every push commit? In WP, there are many commits happening day by day, so instead of measuring it every time, can we measure it once for each release and compare the results to every other commit?

**Comment by youknowriad** [comment]
I've discussed baseMetrics with @joemcgill a few days ago and I'm sharing here part of that discussion for clarity.

We do rely on Github CI Runner to run these tests. That runner CPU, memory... is not stable over time, its OS is not stable over time as Github has to maintain these and make changes to runners over time. (Even if we had our own server, it will be impossible to keep the same conditions for a long period of time). This means that if you run the same test for the same code twice, you'll get different results which also means that we can't compare numbers across different job executions. For this reason, we're forced to compare to a given fixed mark for each job. This allows us to do cross-job comparison by checking the evolution of the numbers relatively to the fixed mark.

Let me explain:

First log: I send 10 for baseMetric (WP 6.1 example) and 15  for commit 1 ===> this get stored as 10 and 15 (no changes)
Second log: I send 12 for baseMetric (6.1) and 20 for commit 2
commit 2 is stored as 20 * 10 / 12 and not just 20

The assumption we're making is that the real performance won’t change for the base Metric (because it’s the exact same code I’m testing) but the environment may change (github runner...) so we normalize it

This is not a perfect solution but it’s the best I’ve had so far with our available resources.

**Comment by joemcgill** [comment]
Thanks, @youknowriad. That explanation helps give more context for why we're measuring a set of baseMetrics during each run. I think it's worth a longer term conversation about whether normalizing data before recording them versus saving the actual numbers measured and normalizing them in the reporting is something that would be valuable, but this makes sense to me for now.

**Comment by felixarntz** [comment]
@youknowriad Thank you for the additional context, I like that idea a lot actually. Even though it's impossible to get 100% clean metrics, this approach definitely should get us closer to that.

One suggestion I would have as a potential enhancement would be to store all values though: I think it's good to also keep the _actual_ measured value in the database, not just the one based on the calculation. Having the raw data is always a good idea in case we later find some mistake or we simply want to do something with the raw (not normalized) data. So I would suggest to always store:
* Normalized value (like today)
* Raw measured value
* Base metric value measured

We could alternatively not even store the normalized value and always compute it on the fly, but I'm afraid that would be more resource heavy and slower, so it's probably best to still store it even if technically it is redundant to store. Let me know what you think about this, we can also discuss in an issue on your Code Health repo.

Another thing that I talked about with both @joemcgill and @oandregal is that (I think) in the Gutenberg repository you always change the base metric when a new WordPress core version is released to be based on that version. After discussing and thinking about this, I think this makes sense, but it just means it's important to keep in mind that the dashboard therefore must not be used to compare performance long-term - since if the base metric changes, the normalization changes, so the data is no longer neutral. I think that is perfectly okay, since the main purpose of the dashboard is probably to measure more short-term performance changes within a WP core release cycle, but we have to be aware of that. I would assume that by default someone who looks at the dashboard would not know about that caveat.

That last point would actually be another reason why it could be useful to additionally store the raw metrics. If we ever want to do something with them, we could display them raw, which would give us more unstable data in the short term, but at the same time the data would be more accurately represent long term performance across WP core releases.

Sorry for hijacking this issue, definitely happy to move this to the Code Health repo. :)

**Comment by joemcgill** [comment]
I've created [a draft PR](https://github.com/10up/wordpress-develop/pull/36) that handles this by using WP CLI to install the specific WP version we want to use for the `baseMetrics` but I got stuck trying to figure out why the docker environment is getting set up using the `src` directory for WP instead of the `build` directory, which is what I would expect given that we're setting 'build' as the `LOCAL_DIR` environment variable.

@eclarke1 I've also added an estimate to this issue, per your request, after chatting with @mukeshpanchal27 earlier today.

**Comment by youknowriad** [comment]
> We could alternatively not even store the normalized value and always compute it on the fly, but I'm afraid that would be more resource heavy and slower, so it's probably best to still store it even if technically it is redundant to store. Let me know what you think about this, we can also discuss in an issue on your Code Health repo.

That's a valid idea, more data is better than no data. Feel free to open an issue on CodeVitals (renamed) repo to continue this.

> Another thing that I talked about with both @joemcgill and @oandregal is that (I think) in the Gutenberg repository you always change the base metric when a new WordPress core version is released to be based on that version. After discussing and thinking about this, I think this makes sense, but it just means it's important to keep in mind that the dashboard therefore must not be used to compare performance long-term - since if the base metric changes, the normalization changes, so the data is no longer neutral. I think that is perfectly okay, since the main purpose of the dashboard is probably to measure more short-term performance changes within a WP core release cycle, but we have to be aware of that. I would assume that by default someone who looks at the dashboard would not know about that caveat.

Yes, in the ideal scenario, the "base" never changes, the problem is that we can't maintain performance tests running for very long range of versions... I mean we can try, but I'm sure they'll start failing at some point.

Now for long term comparison, I agree with you that the change in the "base" have the potential to introduce a "jump" or something that would be unexpected. I haven't found a reliable solution for that yet but I agree with you that it's not a major problem.

Now for actually showing the raw data without any normalization, I actually think that's meaningless as in my experience, I've seen big variations over time between job runs.

**Comment by joemcgill** [comment]
Resolved by #36
